samples/tutorial/Tutorial4.md

```diff
@@ -0,0 +1,343 @@
+# Step 4
+
+_Refactoring and rebalancing a tree of Workflows_
+
+## Setup
+
+To follow this tutorial, launch Android Studio and open this folder (`samples/tutorial`).
+
+Start from the implementation of `tutorial-3-complete` if you're skipping ahead.
+
+## Adding new todo items
+
+A gap in the usability of the todo app is that it does not let the user create new todo items. We will add an "add" button on the right side of the navigation bar for this.
+
+## Refactoring a workflow by splitting it into a parent and child
+
+The `TodoListWorkflow` has started to grow and has multiple concerns it's handling — specifically all of the `TutorialListScreen` behavior, as well as the actions that can come from the `TodoEditWorkflow`.
+
+When a single workflow seems to be doing too many things, a common pattern is to extract some of its responsibility into a parent.
+
+### TodoWorkflow
+
+Create a new workflow called `Todo` that will be responsible for both the `TodoListWorkflow` and  the `TodoEditWorkflow`.
+
+```kotlin
+object TodoWorkflow : StatefulWorkflow<TodoProps, State, Back, List<Any>>() {
+
+  // …
+
+}
+```
+
+#### Moving logic from the TodoList to the TodoWorkflow
+
+Move the `ListState` state, input, and outputs from the `TodoListWorkflow` up to the `TodoWorkflow`. It will be owner the list of todo items, and the `TodoListWorkflow` will simply show whatever is passed into its input:
+
+```kotlin
+object TodoWorkflow : StatefulWorkflow<TodoProps, State, Back, List<Any>>() {
+
+  data class TodoProps(val name: String)
```

Comment:
Cross cutting nit: thinking that `userName` would be clearer for this and the like state fields.

Reply:
#246, gonna do in a follow up so i can change everywhere at once.